### ISSUE TITLE
How to show combinator labels between rules?

### ISSUE THREAD

**Issue by muaragon** [react-querybuilder/react-querybuilder#249]
Hi!
I am trying to display a combinator label between each rule of my querybuilder. My initial approach was to try to override the Rule Component, but I realized that this component does not have access to its left or right sibling combinator.  My intuition tells me that I will have to override the RuleGroup component, but I wonder if there is an easier way to access combinator values?

# Figma design 
![image](https://user-images.githubusercontent.com/86376904/146266778-60cefacc-66af-4f58-be69-6aeab8b0ffdd.png)

**Comment by jakeboone02** [comment]
If you're trying to display the combinator selectors for each group between the rules instead of in the group header, then you can use the `showCombinatorsBetweenRules` prop. That will display a combinator selector between each pair of rules, but the combinator must still be the same within each group (e.g. changing one combinator from 'and' to 'or' will change _all_ the sibling combinators in that group to 'or').

If you actually want the combinators to be independent of each other, then I've got good news for you! `react-querybuilder` v4 will introduce the `independentCombinators` prop, which allows each combinator to be selected independently of all others, even siblings. You can try out the new prop with the 4.0 beta version:

```bash
npm install --save react-querybuilder@4.0.0-beta.7
```

or

```bash
yarn add react-querybuilder@4.0.0-beta.7
```

Documentation for v4.0 is here: https://react-querybuilder.js.org/

**Comment by muaragon** [comment]
Thank you for your response!

I don't want to display a combinator selector between each rule, only a label for the boolean value. I plan to keep the combinator selector at the top level of a group as shown in the updated Figma screenshot. 

Do you think this can only be achieved by overriding the RuleGroup?

**Comment by jakeboone02** [comment]
Do you just want to reorder the elements in the group header, to put the combinator selector on the far right? If so, v4 will be able to help you out there as well. All the default styles in v4 are based on flexbox, so you can reorder elements with CSS.

**Comment by muaragon** [comment]
I want to show ("OR" , "AND") text labels between rules and ruleGroups. To do this, I need access combinator values, but I am not sure what is the best place to extract them. 

![image](https://user-images.githubusercontent.com/86376904/146272192-7789065d-d6b4-4883-94fa-48c6c638812d.png)

**Comment by jakeboone02** [comment]
Oh, I think I see now -- you want the actual combinator selector to be in the group header like normal, but _also_ want the combinator label to show between rules? Hmm...I wouldn't want you to have to override the `RuleGroup` component just for that. Let me think about it a bit.

**Comment by jakeboone02** [comment]
@muaragon - I think I have something like what you want working here:  https://codesandbox.io/s/react-querybuilder-multi-combinator-dfib7?file=/src/App.tsx

It's a little convoluted, but it's definitely better than overriding the entire `RuleGroup` component.

Note: this uses `v4.0.0-beta.7` which internally relies on `path: number[]` instead of `id: string`, but the same can be done with `v3.12.1` if you just replace `path`s with `id`s and use `findRule` instead of `findPath`.

**Comment by jakeboone02** [comment]
Actually, never mind about v3.12.1. The method used in that codesandbox only works in v4 because v3 doesn't pass down the group's `id` to subcomponents (like v4 passes down the `path`).

**Comment by muaragon** [comment]
It looks almost exactly to what I want! 

I notice that the labels are not independent when selecting a combinator in the group header, is it possible to use the independentCombinators flag for this logic as well?
Thanks!

**Comment by jakeboone02** [comment]
If you specified `independentCombinators`, you'd have to decide what the behavior would be. What would you want to happen if the combinator selector in the group header was changed? Would it reset all of that group's independent combinators to the same value? What value would the group header combinator selector display if you had multiple values for independent combinators within that group? Also, you'd need the independent combinator selector to _actually_ be a selector, not just static label text.

In short, yes you could do it but you would need to make some UX decisions.

**Comment by muaragon** [comment]
I don't want the combinator selector in the group header to reset any of the previous combinator. I only want it to alter new combinators. Do you think this is possible?

Thank you for all of your help so far!

**Comment by jakeboone02** [comment]
Like this?  https://codesandbox.io/s/react-querybuilder-default-independent-combinators-fyvki?file=/src/App.tsx

**Comment by muaragon** [comment]
Yes!
I added the "CombinatorLabel" that you provided me in the previous Sandbox and this is the exact functionality that I was looking for. I really appreciate your help and thank you for creating this library!

**Comment by jakeboone02** [comment]
Great! Glad to help. (BTW I'm not the original creator, just the lone remaining maintainer.)
